async keyword overly sensitive in VS2015

When I write a method that returns `Task` and apply thr `async` modifier, then the intellisense in VS2015 seems to have lots of problems. If anything in the method doesn't compile, then the intellisense is broken. This is often due to something later in the method with a syntax error – it causes intellisense for everything prior in the method to just now work (this used to be ok in VS2013).

This is not going to be a simple fix, I think. 
The fact that C# compiler is unable to combine both instance and extension methods within a single ```Microsoft.CodeAnalysis.CSharp.MethodGroupResolution``` is at the root of the issue. Given the same call site, AnalyzedArguments stored inside MethodGroupResolution are going to be different for a group of instance methods and for a group of extension methods. I think for extension methods it includes the receiver as well. 
This creates a problem when we have candidates of both kinds and none of them are applicable. Since the candidates of different kind cannot be merged into a single MethodGroupResolution, Microsoft.CodeAnalysis.CSharp.Binder.ResolveMethodGroupInternal have to choose one or another and, when none of the candidates are applicable, it prefers to keep instance methods. The information about presence of extension methods candidates is lost after that point and the rest of the binding pipeline operates as if they do not exist. 
```
Microsoft.CodeAnalysis.CSharp.Binder
        private MethodGroupResolution ResolveMethodGroupInternal(
            BoundMethodGroup methodGroup,
            CSharpSyntaxNode expression,
            string methodName,
            AnalyzedArguments analyzedArguments,
            bool isMethodGroupConversion,
            ref HashSet<DiagnosticInfo> useSiteDiagnostics,
            bool inferWithDynamic = false,
            bool allowUnexpandedForm = true)
```

For our original repro scenario (this is GetMemberGroupInsideIncompleteLambda_01 unit-test, see commit https://github.com/AlekseyTs/roslyn/commit/1d2d8ad790e0c344fa0e7b1726417e1a21f64a85), for the ```app.Use``` invocation we end up with MethodGroupResolution containing only one method , ```IApplicationBuilder IApplicationBuilder.Use(System.Func<RequestDelegate, RequestDelegate> middleware)```, which is an instance method. The other candidate, ```IApplicationBuilder IApplicationBuilderExtensions.Use(this IApplicationBuilder app, System.Func<HttpContext, System.Func<System.Threading.Tasks.Task>, System.Threading.Tasks.Task> middleware)```, is getting dropped due to the reason described earlier (at this point, it is also considered inapplicable because the lambda is incomplete and, therefore, has errors). This fact can be observed by examining result returned by SemanticMode.GetSymbolInfo API for ```app.Use```, CandidateSymbols consists of only one symbol, which represents the instance method candidate. In contrast, if both methods were instance methods, or both methods were extension methods, both of them would be in CandidateSymbols set (see GetMemberGroupInsideIncompleteLambda_03 and GetMemberGroupInsideIncompleteLambda_04 unit-test).

Eventually, Microsoft.CodeAnalysis.CSharp.Binder.CreateBadCall is getting called to create a bound node for the ```app.Use``` invocation.
```
Microsoft.CodeAnalysis.CSharp.Binder
        private BoundCall CreateBadCall(
            CSharpSyntaxNode node,
            string name,
            BoundExpression receiver,
            ImmutableArray<MethodSymbol> methods,
            LookupResultKind resultKind,
            ImmutableArray<TypeSymbol> typeArguments,
            AnalyzedArguments analyzedArguments,
            bool invokedAsExtensionMethod,
            bool isDelegate)
```
It has a special logic, if there is only one candidate method, all arguments are strongly tied to that candidate. I.e. bound node for the lambda will be a BoundLambda bound against target delegate type for that single candidate. In our case the target delegate type is System.Func<RequestDelegate, RequestDelegate>, which doesn’t match the shape of the lambda, therefore, no meaningful semantic information can be provided for the body of the lambda. 

However, the diagnostics reported for the invocation ``` ctx.Authentication.AuthenticateAsync```:
```
error CS7036: There is no argument given that corresponds to the required formal parameter 'authenticationScheme' of 'AuthenticationManager.AuthenticateAsync(string)'
   await ctx.Authentication.AuthenticateAsync();
```
might lead an observer into thinking that compiler actually considered extension method as a “better” inapplicable candidate. That is not the case. There is a logic in Microsoft.CodeAnalysis.CSharp.OverloadResolutionResult<TMember>.ReportDiagnostics<T>
```
Microsoft.CodeAnalysis.CSharp.OverloadResolutionResult<TMember>
        internal void ReportDiagnostics<T>(
            Binder binder,
            Location location,
            DiagnosticBag diagnostics,
            string name,
            BoundExpression receiver,
            AnalyzedArguments arguments,
            ImmutableArray<T> memberGroup, // the T is just a convenience for the caller
            NamedTypeSymbol typeContainingConstructor,
            NamedTypeSymbol delegateTypeBeingInvoked,
            CSharpSyntaxNode queryClause = null,
            bool isMethodGroupConversion = false) where T : Symbol
```
to report errors from lambda bodies, if there are any lambdas among the arguments. And prioritize these errors over other argument errors. The errors are collected across all trial bindings performed for a given lambda. That is why an error related to a candidate dropped by ResolveMethodGroupInternal makes its way to the top, even though it is not even related to the method compiler has.

I think that the proper way to address the issue is to be able to combine candidates of different kinds in a single MethodGroupResolution without having to make an arbitrary choice between instance and extension methods.

--- a/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
@@ -2771,7 +2771,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                         typeArguments: ImmutableArray<TypeSymbol>.Empty,
                         analyzedArguments: analyzedArguments,
                         invokedAsExtensionMethod: false,
-                        isDelegate: false);
+                        isDelegate: false,
+                        extensionMethodsOfSameViabilityAreAvailable: false);
                     result.WasCompilerGenerated = initializerArgumentListOpt == null;
                     return result;
                 }
@@ -5173,7 +5174,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                 OverloadResolution.MethodInvocationOverloadResolution(methodGroup.Methods, methodGroup.TypeArguments, actualArguments, overloadResolutionResult, ref useSiteDiagnostics, isMethodGroupConversion, allowRefOmittedArguments);
                 diagnostics.Add(expression, useSiteDiagnostics);
                 var sealedDiagnostics = diagnostics.ToReadOnlyAndFree();
-                var result = new MethodGroupResolution(methodGroup, null, overloadResolutionResult, actualArguments, methodGroup.ResultKind, sealedDiagnostics);
+                var result = new MethodGroupResolution(methodGroup, null, overloadResolutionResult, actualArguments, methodGroup.ResultKind, sealedDiagnostics,
+                                                       extensionMethodsOfSameViabilityAreAvailable: false);
 
                 // If the search in the current scope resulted in any applicable method (regardless of whether a best 
                 // applicable method could be determined) then our search is complete. Otherwise, store aside the
@@ -6167,7 +6169,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                 var diagnostics = DiagnosticBag.GetInstance();
                 diagnostics.AddRange(methodResolution.Diagnostics); // Could still have use site warnings.
                 BindMemberAccessReportError(node, diagnostics);
-                return new MethodGroupResolution(methodResolution.MethodGroup, methodResolution.OtherSymbol, methodResolution.OverloadResolutionResult, methodResolution.AnalyzedArguments, methodResolution.ResultKind, diagnostics.ToReadOnlyAndFree());
+                return new MethodGroupResolution(methodResolution.MethodGroup, methodResolution.OtherSymbol, methodResolution.OverloadResolutionResult, methodResolution.AnalyzedArguments, methodResolution.ResultKind, diagnostics.ToReadOnlyAndFree(),
+                                                 methodResolution.ExtensionMethodsOfSameViabilityAreAvailable);
             }
             return methodResolution;
         }
@@ -6196,6 +6199,8 @@ namespace Microsoft.CodeAnalysis.CSharp
             var extensionMethodResolution = BindExtensionMethod(expression, methodName, analyzedArguments, methodGroup.ReceiverOpt, methodGroup.TypeArgumentsOpt, isMethodGroupConversion);
 
             bool preferExtensionMethodResolution = false;
+            bool extensionMethodsOfSameViabilityAreAvailable = false;
+
             if (extensionMethodResolution.HasAnyApplicableMethod)
             {
                 preferExtensionMethodResolution = true;
@@ -6220,8 +6225,12 @@ namespace Microsoft.CodeAnalysis.CSharp
 
                 LookupResultKind methodResultKind = methodResolution.ResultKind;
                 LookupResultKind extensionMethodResultKind = extensionMethodResolution.ResultKind;
-                if (methodResultKind != extensionMethodResultKind &&
-                    methodResultKind == extensionMethodResultKind.WorseResultKind(methodResultKind))
+                if (methodResultKind == extensionMethodResultKind)
+                {
+                    // This information allows us to preserve more useful information for SemanticModel.
+                    extensionMethodsOfSameViabilityAreAvailable = true;
+                }
+                else if (methodResultKind == extensionMethodResultKind.WorseResultKind(methodResultKind))
                 {
                     preferExtensionMethodResolution = true;
                 }
@@ -6235,6 +6244,18 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
 
             extensionMethodResolution.Free();
+
+            if (extensionMethodsOfSameViabilityAreAvailable)
+            {
+                return new MethodGroupResolution(methodResolution.MethodGroup,
+                                                 methodResolution.OtherSymbol,
+                                                 methodResolution.OverloadResolutionResult, 
+                                                 methodResolution.AnalyzedArguments, 
+                                                 methodResolution.ResultKind,
+                                                 methodResolution.Diagnostics,
+                                                 extensionMethodsOfSameViabilityAreAvailable: true);
+            }
+
             return methodResolution;
         }
 
@@ -6293,7 +6314,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                     methodGroup.Methods, methodGroup.TypeArguments, analyzedArguments,
                     result, ref useSiteDiagnostics, isMethodGroupConversion, allowRefOmittedArguments,
                     inferWithDynamic: inferWithDynamic, allowUnexpandedForm: allowUnexpandedForm);
-                return new MethodGroupResolution(methodGroup, null, result, analyzedArguments, methodGroup.ResultKind, sealedDiagnostics);
+                return new MethodGroupResolution(methodGroup, null, result, analyzedArguments, methodGroup.ResultKind, sealedDiagnostics,
+                                                 extensionMethodsOfSameViabilityAreAvailable: false);
             }
         }
 

--- a/src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs
@@ -415,7 +415,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             }
             else
             {
-                result = BindInvocationExpressionContinued(node, expression, methodName, overloadResolutionResult, analyzedArguments, methodGroup, delegateType, diagnostics, queryClause);
+                result = BindInvocationExpressionContinued(node, expression, methodName, overloadResolutionResult, analyzedArguments, methodGroup, delegateType, diagnostics, false, queryClause);
             }
 
             overloadResolutionResult.Free();
@@ -483,7 +483,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                     typeArguments,
                     analyzedArguments,
                     invokedAsExtensionMethod: resolution.IsExtensionMethodGroup,
-                    isDelegate: false);
+                    isDelegate: false,
+                    extensionMethodsOfSameViabilityAreAvailable: resolution.ExtensionMethodsOfSameViabilityAreAvailable);
             }
             else if (!resolution.IsEmpty)
             {
@@ -498,7 +499,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                     {
                         // we want to force any unbound lambda arguments to cache an appropriate conversion if possible; see 9448.
                         DiagnosticBag discarded = DiagnosticBag.GetInstance();
-                        result = BindInvocationExpressionContinued(syntax, expression, methodName, resolution.OverloadResolutionResult, resolution.AnalyzedArguments, resolution.MethodGroup, null, discarded, queryClause);
+                        result = BindInvocationExpressionContinued(syntax, expression, methodName, resolution.OverloadResolutionResult, resolution.AnalyzedArguments, resolution.MethodGroup, null, discarded, 
+                                                                   resolution.ExtensionMethodsOfSameViabilityAreAvailable, queryClause);
                         discarded.Free();
                     }
 
@@ -553,7 +555,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                     {
                         result = BindInvocationExpressionContinued(
                             syntax, expression, methodName, resolution.OverloadResolutionResult, resolution.AnalyzedArguments,
-                            resolution.MethodGroup, null, diagnostics, queryClause);
+                            resolution.MethodGroup, null, diagnostics, resolution.ExtensionMethodsOfSameViabilityAreAvailable, queryClause);
                     }
                 }
             }
@@ -676,6 +678,10 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// <param name="methodGroup">Method group if the invocation represents a potentially overloaded member.</param>
         /// <param name="delegateTypeOpt">Delegate type if method group represents a delegate.</param>
         /// <param name="diagnostics">Diagnostics.</param>
+        /// <param name="extensionMethodsOfSameViabilityAreAvailable">
+        /// Indicates that there are additional extension method candidates of the same lookup viability in cases when 
+        /// none of the instance methods are applicable to the argument list.
+        /// </param>
         /// <param name="queryClause">The syntax for the query clause generating this invocation expression, if any.</param>
         /// <returns>BoundCall or error expression representing the invocation.</returns>
         private BoundCall BindInvocationExpressionContinued(
@@ -687,6 +693,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             MethodGroup methodGroup,
             NamedTypeSymbol delegateTypeOpt,
             DiagnosticBag diagnostics,
+            bool extensionMethodsOfSameViabilityAreAvailable,
             CSharpSyntaxNode queryClause = null)
         {
             Debug.Assert(node != null);
@@ -726,7 +733,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                 }
 
                 return CreateBadCall(node, methodGroup.Name, invokedAsExtensionMethod && analyzedArguments.Arguments.Count > 0 && (object)methodGroup.Receiver == (object)analyzedArguments.Arguments[0] ? null : methodGroup.Receiver,
-                    GetOriginalMethods(result), methodGroup.ResultKind, methodGroup.TypeArguments.ToImmutable(), analyzedArguments, invokedAsExtensionMethod: invokedAsExtensionMethod, isDelegate: ((object)delegateTypeOpt != null));
+                    GetOriginalMethods(result), methodGroup.ResultKind, methodGroup.TypeArguments.ToImmutable(), analyzedArguments, invokedAsExtensionMethod: invokedAsExtensionMethod, isDelegate: ((object)delegateTypeOpt != null),
+                    extensionMethodsOfSameViabilityAreAvailable: extensionMethodsOfSameViabilityAreAvailable);
             }
 
             // Otherwise, there were no dynamic arguments and overload resolution found a unique best candidate. 
@@ -950,7 +958,8 @@ namespace Microsoft.CodeAnalysis.CSharp
             ImmutableArray<TypeSymbol> typeArguments,
             AnalyzedArguments analyzedArguments,
             bool invokedAsExtensionMethod,
-            bool isDelegate)
+            bool isDelegate,
+            bool extensionMethodsOfSameViabilityAreAvailable)
         {
             MethodSymbol method;
             ImmutableArray<BoundExpression> args;
@@ -965,10 +974,11 @@ namespace Microsoft.CodeAnalysis.CSharp
                 methods = constructedMethods.ToImmutableAndFree();
             }
 
-            if (methods.Length == 1)
+            if (methods.Length == 1 && !extensionMethodsOfSameViabilityAreAvailable)
             {
-                // If there is only one method in the group, we should attempt to bind to it.  That includes
-                // binding any lambdas in the argument list against the method's parameter types.
+                // If there is only one method in the group and no additional extension methods with the same lookup viability,
+                // we should attempt to bind to it.  That includes binding any lambdas in the argument list against
+                // the method's parameter types.
                 method = methods[0];
                 args = BuildArgumentsForErrorRecovery(analyzedArguments, method.Parameters);
             }

--- a/src/Compilers/CSharp/Portable/Binder/MethodGroupResolution.cs
+++ b/src/Compilers/CSharp/Portable/Binder/MethodGroupResolution.cs
@@ -19,9 +19,10 @@ namespace Microsoft.CodeAnalysis.CSharp
         public readonly AnalyzedArguments AnalyzedArguments;
         public readonly ImmutableArray<Diagnostic> Diagnostics;
         public readonly LookupResultKind ResultKind;
+        public readonly bool ExtensionMethodsOfSameViabilityAreAvailable;
 
         public MethodGroupResolution(MethodGroup methodGroup, ImmutableArray<Diagnostic> diagnostics)
-            : this(methodGroup, null, null, null, methodGroup.ResultKind, diagnostics)
+            : this(methodGroup, null, null, null, methodGroup.ResultKind, diagnostics, false)
         {
         }
 
@@ -30,12 +31,12 @@ namespace Microsoft.CodeAnalysis.CSharp
             OverloadResolutionResult<MethodSymbol> overloadResolutionResult,
             AnalyzedArguments analyzedArguments,
             ImmutableArray<Diagnostic> diagnostics)
-            : this(methodGroup, null, overloadResolutionResult, analyzedArguments, methodGroup.ResultKind, diagnostics)
+            : this(methodGroup, null, overloadResolutionResult, analyzedArguments, methodGroup.ResultKind, diagnostics, false)
         {
         }
 
         public MethodGroupResolution(Symbol otherSymbol, LookupResultKind resultKind, ImmutableArray<Diagnostic> diagnostics)
-            : this(null, otherSymbol, null, null, resultKind, diagnostics)
+            : this(null, otherSymbol, null, null, resultKind, diagnostics, false)
         {
         }
 
@@ -45,7 +46,8 @@ namespace Microsoft.CodeAnalysis.CSharp
             OverloadResolutionResult<MethodSymbol> overloadResolutionResult,
             AnalyzedArguments analyzedArguments,
             LookupResultKind resultKind,
-            ImmutableArray<Diagnostic> diagnostics)
+            ImmutableArray<Diagnostic> diagnostics,
+            bool extensionMethodsOfSameViabilityAreAvailable)
         {
             Debug.Assert((methodGroup == null) || (methodGroup.Methods.Count > 0));
             Debug.Assert((methodGroup == null) || ((object)otherSymbol == null));
@@ -53,6 +55,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             Debug.Assert(((object)otherSymbol == null) || (otherSymbol.Kind != SymbolKind.Method));
             Debug.Assert(resultKind != LookupResultKind.Ambiguous); // HasAnyApplicableMethod is expecting Viable methods.
             Debug.Assert(!diagnostics.IsDefault);
+            Debug.Assert(!extensionMethodsOfSameViabilityAreAvailable || methodGroup == null || !methodGroup.IsExtensionMethodGroup);
 
             this.MethodGroup = methodGroup;
             this.OtherSymbol = otherSymbol;
@@ -60,6 +63,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             this.AnalyzedArguments = analyzedArguments;
             this.ResultKind = resultKind;
             this.Diagnostics = diagnostics;
+            this.ExtensionMethodsOfSameViabilityAreAvailable = extensionMethodsOfSameViabilityAreAvailable;
         }
 
         public bool IsEmpty

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/BindingTests.cs
@@ -2727,5 +2727,324 @@ public static class LazyToStringExtension
             var param = node.ChildNodes().Where(n => n.IsKind(SyntaxKind.Parameter)).Single();
             Assert.Equal("System.Reflection.PropertyInfo x", model.GetDeclaredSymbol(param).ToTestDisplayString());
         }
+
+        [Fact, WorkItem(5128, "https://github.com/dotnet/roslyn/issues/5128")]
+        public void GetMemberGroupInsideIncompleteLambda_01()
+        {
+            var source =
+@"
+using System;
+using System.Threading.Tasks;
+
+public delegate Task RequestDelegate(HttpContext context);
+
+public class AuthenticationResult { }
+
+public abstract class AuthenticationManager
+{
+    public abstract Task<AuthenticationResult> AuthenticateAsync(string authenticationScheme);
+}
+
+public abstract class HttpContext
+{
+    public abstract AuthenticationManager Authentication { get; }
+}
+
+interface IApplicationBuilder
+{
+    IApplicationBuilder Use(Func<RequestDelegate, RequestDelegate> middleware);
+}
+
+static class IApplicationBuilderExtensions
+{
+    public static IApplicationBuilder Use(this IApplicationBuilder app, Func<HttpContext, Func<Task>, Task> middleware)
+    {
+        return app;
+    }
+}
+
+class C
+{
+    void M(IApplicationBuilder app)
+    {
+        app.Use(async (ctx, next) =>
+        {
+            await ctx.Authentication.AuthenticateAsync();
+        });
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, new[] { SystemCoreRef });
+
+            comp.VerifyDiagnostics(
+    // (41,38): error CS7036: There is no argument given that corresponds to the required formal parameter 'authenticationScheme' of 'AuthenticationManager.AuthenticateAsync(string)'
+    //             await ctx.Authentication.AuthenticateAsync();
+    Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "AuthenticateAsync").WithArguments("authenticationScheme", "AuthenticationManager.AuthenticateAsync(string)").WithLocation(38, 38)
+                );
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var node1 = tree.GetRoot().DescendantNodes().Where(n => n.IsKind(SyntaxKind.IdentifierName) && ((IdentifierNameSyntax)n).Identifier.ValueText == "Use").Single().Parent;
+            Assert.Equal("app.Use", node1.ToString());
+            var group1 = model.GetMemberGroup(node1);
+            Assert.Equal(2, group1.Length);
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<RequestDelegate, RequestDelegate> middleware)", group1[0].ToTestDisplayString());
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<HttpContext, System.Func<System.Threading.Tasks.Task>, System.Threading.Tasks.Task> middleware)", 
+                         group1[1].ToTestDisplayString());
+
+            var symbolInfo1 = model.GetSymbolInfo(node1);
+            Assert.Null(symbolInfo1.Symbol);
+            Assert.Equal(1, symbolInfo1.CandidateSymbols.Length);
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<RequestDelegate, RequestDelegate> middleware)", symbolInfo1.CandidateSymbols.Single().ToTestDisplayString());
+            Assert.Equal(CandidateReason.OverloadResolutionFailure, symbolInfo1.CandidateReason);
+
+            var node = tree.GetRoot().DescendantNodes().Where(n => n.IsKind(SyntaxKind.IdentifierName) && ((IdentifierNameSyntax)n).Identifier.ValueText== "AuthenticateAsync").Single().Parent;
+
+            Assert.Equal("ctx.Authentication.AuthenticateAsync", node.ToString());
+
+            var group = model.GetMemberGroup(node);
+
+            Assert.Equal("System.Threading.Tasks.Task<AuthenticationResult> AuthenticationManager.AuthenticateAsync(System.String authenticationScheme)", group.Single().ToTestDisplayString());
+        }
+
+        [Fact, WorkItem(5128, "https://github.com/dotnet/roslyn/issues/5128")]
+        public void GetMemberGroupInsideIncompleteLambda_02()
+        {
+            var source =
+@"
+using System;
+using System.Threading.Tasks;
+
+public delegate Task RequestDelegate(HttpContext context);
+
+public class AuthenticationResult { }
+
+public abstract class AuthenticationManager
+{
+    public abstract Task<AuthenticationResult> AuthenticateAsync(string authenticationScheme);
+}
+
+public abstract class HttpContext
+{
+    public abstract AuthenticationManager Authentication { get; }
+}
+
+interface IApplicationBuilder
+{
+    IApplicationBuilder Use(Func<HttpContext, Func<Task>, Task> middleware);
+}
+
+static class IApplicationBuilderExtensions
+{
+    public static IApplicationBuilder Use(this IApplicationBuilder app, Func<RequestDelegate, RequestDelegate> middleware)
+    {
+        return app;
+    }
+}
+
+class C
+{
+    void M(IApplicationBuilder app)
+    {
+        app.Use(async (ctx, next) =>
+        {
+            await ctx.Authentication.AuthenticateAsync();
+        });
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, new[] { SystemCoreRef });
+
+            comp.VerifyDiagnostics(
+    // (41,38): error CS7036: There is no argument given that corresponds to the required formal parameter 'authenticationScheme' of 'AuthenticationManager.AuthenticateAsync(string)'
+    //             await ctx.Authentication.AuthenticateAsync();
+    Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "AuthenticateAsync").WithArguments("authenticationScheme", "AuthenticationManager.AuthenticateAsync(string)").WithLocation(38, 38)
+                );
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var node1 = tree.GetRoot().DescendantNodes().Where(n => n.IsKind(SyntaxKind.IdentifierName) && ((IdentifierNameSyntax)n).Identifier.ValueText == "Use").Single().Parent;
+            Assert.Equal("app.Use", node1.ToString());
+            var group1 = model.GetMemberGroup(node1);
+            Assert.Equal(2, group1.Length);
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<HttpContext, System.Func<System.Threading.Tasks.Task>, System.Threading.Tasks.Task> middleware)", 
+                         group1[0].ToTestDisplayString());
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<RequestDelegate, RequestDelegate> middleware)", group1[1].ToTestDisplayString());
+
+            var symbolInfo1 = model.GetSymbolInfo(node1);
+            Assert.Null(symbolInfo1.Symbol);
+            Assert.Equal(1, symbolInfo1.CandidateSymbols.Length);
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<HttpContext, System.Func<System.Threading.Tasks.Task>, System.Threading.Tasks.Task> middleware)", symbolInfo1.CandidateSymbols.Single().ToTestDisplayString());
+            Assert.Equal(CandidateReason.OverloadResolutionFailure, symbolInfo1.CandidateReason);
+
+            var node = tree.GetRoot().DescendantNodes().Where(n => n.IsKind(SyntaxKind.IdentifierName) && ((IdentifierNameSyntax)n).Identifier.ValueText == "AuthenticateAsync").Single().Parent;
+
+            Assert.Equal("ctx.Authentication.AuthenticateAsync", node.ToString());
+
+            var group = model.GetMemberGroup(node);
+
+            Assert.Equal("System.Threading.Tasks.Task<AuthenticationResult> AuthenticationManager.AuthenticateAsync(System.String authenticationScheme)", group.Single().ToTestDisplayString());
+        }
+
+        [Fact, WorkItem(5128, "https://github.com/dotnet/roslyn/issues/5128")]
+        public void GetMemberGroupInsideIncompleteLambda_03()
+        {
+            var source =
+@"
+using System;
+using System.Threading.Tasks;
+
+public delegate Task RequestDelegate(HttpContext context);
+
+public class AuthenticationResult { }
+
+public abstract class AuthenticationManager
+{
+    public abstract Task<AuthenticationResult> AuthenticateAsync(string authenticationScheme);
+}
+
+public abstract class HttpContext
+{
+    public abstract AuthenticationManager Authentication { get; }
+}
+
+interface IApplicationBuilder
+{
+    IApplicationBuilder Use(Func<RequestDelegate, RequestDelegate> middleware);
+    IApplicationBuilder Use(Func<HttpContext, Func<Task>, Task> middleware);
+}
+
+class C
+{
+    void M(IApplicationBuilder app)
+    {
+        app.Use(async (ctx, next) =>
+        {
+            await ctx.Authentication.AuthenticateAsync();
+        });
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, new[] { SystemCoreRef });
+
+            comp.VerifyDiagnostics(
+    // (41,38): error CS7036: There is no argument given that corresponds to the required formal parameter 'authenticationScheme' of 'AuthenticationManager.AuthenticateAsync(string)'
+    //             await ctx.Authentication.AuthenticateAsync();
+    Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "AuthenticateAsync").WithArguments("authenticationScheme", "AuthenticationManager.AuthenticateAsync(string)").WithLocation(31, 38)
+                );
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var node1 = tree.GetRoot().DescendantNodes().Where(n => n.IsKind(SyntaxKind.IdentifierName) && ((IdentifierNameSyntax)n).Identifier.ValueText == "Use").Single().Parent;
+            Assert.Equal("app.Use", node1.ToString());
+            var group1 = model.GetMemberGroup(node1);
+            Assert.Equal(2, group1.Length);
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<RequestDelegate, RequestDelegate> middleware)", group1[0].ToTestDisplayString());
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<HttpContext, System.Func<System.Threading.Tasks.Task>, System.Threading.Tasks.Task> middleware)", 
+                         group1[1].ToTestDisplayString());
+
+            var symbolInfo1 = model.GetSymbolInfo(node1);
+            Assert.Null(symbolInfo1.Symbol);
+            Assert.Equal(2, symbolInfo1.CandidateSymbols.Length);
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<RequestDelegate, RequestDelegate> middleware)", symbolInfo1.CandidateSymbols[0].ToTestDisplayString());
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<HttpContext, System.Func<System.Threading.Tasks.Task>, System.Threading.Tasks.Task> middleware)", symbolInfo1.CandidateSymbols[1].ToTestDisplayString());
+            Assert.Equal(CandidateReason.OverloadResolutionFailure, symbolInfo1.CandidateReason);
+
+            var node = tree.GetRoot().DescendantNodes().Where(n => n.IsKind(SyntaxKind.IdentifierName) && ((IdentifierNameSyntax)n).Identifier.ValueText == "AuthenticateAsync").Single().Parent;
+
+            Assert.Equal("ctx.Authentication.AuthenticateAsync", node.ToString());
+
+            var group = model.GetMemberGroup(node);
+
+            Assert.Equal("System.Threading.Tasks.Task<AuthenticationResult> AuthenticationManager.AuthenticateAsync(System.String authenticationScheme)", group.Single().ToTestDisplayString());
+        }
+
+        [Fact, WorkItem(5128, "https://github.com/dotnet/roslyn/issues/5128")]
+        public void GetMemberGroupInsideIncompleteLambda_04()
+        {
+            var source =
+@"
+using System;
+using System.Threading.Tasks;
+
+public delegate Task RequestDelegate(HttpContext context);
+
+public class AuthenticationResult { }
+
+public abstract class AuthenticationManager
+{
+    public abstract Task<AuthenticationResult> AuthenticateAsync(string authenticationScheme);
+}
+
+public abstract class HttpContext
+{
+    public abstract AuthenticationManager Authentication { get; }
+}
+
+interface IApplicationBuilder
+{
+}
+
+static class IApplicationBuilderExtensions
+{
+    public static IApplicationBuilder Use(this IApplicationBuilder app, Func<RequestDelegate, RequestDelegate> middleware)
+    {
+        return app;
+    }
+
+    public static IApplicationBuilder Use(this IApplicationBuilder app, Func<HttpContext, Func<Task>, Task> middleware)
+    {
+        return app;
+    }
+}
+
+class C
+{
+    void M(IApplicationBuilder app)
+    {
+        app.Use(async (ctx, next) =>
+        {
+            await ctx.Authentication.AuthenticateAsync();
+        });
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib(source, new[] { SystemCoreRef });
+
+            comp.VerifyDiagnostics(
+    // (41,38): error CS7036: There is no argument given that corresponds to the required formal parameter 'authenticationScheme' of 'AuthenticationManager.AuthenticateAsync(string)'
+    //             await ctx.Authentication.AuthenticateAsync();
+    Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "AuthenticateAsync").WithArguments("authenticationScheme", "AuthenticationManager.AuthenticateAsync(string)").WithLocation(42, 38)
+                );
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var node1 = tree.GetRoot().DescendantNodes().Where(n => n.IsKind(SyntaxKind.IdentifierName) && ((IdentifierNameSyntax)n).Identifier.ValueText == "Use").Single().Parent;
+            Assert.Equal("app.Use", node1.ToString());
+            var group1 = model.GetMemberGroup(node1);
+            Assert.Equal(2, group1.Length);
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<RequestDelegate, RequestDelegate> middleware)", group1[0].ToTestDisplayString());
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<HttpContext, System.Func<System.Threading.Tasks.Task>, System.Threading.Tasks.Task> middleware)", 
+                         group1[1].ToTestDisplayString());
+
+            var symbolInfo1 = model.GetSymbolInfo(node1);
+            Assert.Null(symbolInfo1.Symbol);
+            Assert.Equal(2, symbolInfo1.CandidateSymbols.Length);
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<RequestDelegate, RequestDelegate> middleware)", symbolInfo1.CandidateSymbols[0].ToTestDisplayString());
+            Assert.Equal("IApplicationBuilder IApplicationBuilder.Use(System.Func<HttpContext, System.Func<System.Threading.Tasks.Task>, System.Threading.Tasks.Task> middleware)", symbolInfo1.CandidateSymbols[1].ToTestDisplayString());
+            Assert.Equal(CandidateReason.OverloadResolutionFailure, symbolInfo1.CandidateReason);
+
+            var node = tree.GetRoot().DescendantNodes().Where(n => n.IsKind(SyntaxKind.IdentifierName) && ((IdentifierNameSyntax)n).Identifier.ValueText == "AuthenticateAsync").Single().Parent;
+
+            Assert.Equal("ctx.Authentication.AuthenticateAsync", node.ToString());
+
+            var group = model.GetMemberGroup(node);
+
+            Assert.Equal("System.Threading.Tasks.Task<AuthenticationResult> AuthenticationManager.AuthenticateAsync(System.String authenticationScheme)", group.Single().ToTestDisplayString());
+        }
     }
 }

